### PR TITLE
fix(ios/engine): adds check for keyboard load success, auto-reset on load failure

### LINKED ISSUE #3985: bug(ios): keyboard search + Keyman as default system keyboard loading issues
**Describe the bug**
If Keyman is currently serving as the system-level default keyboard on the device, it seems to have initialization issues when a user first attempts to provide input within our in-app keyboard search.

Screen Shot 2020-11-27 at 1 36 11 PM

This only seems to occur when _**both**_ of the following conditions are true:
- The keyboard is being loaded upon the _first_ selection / focus of the search element.
  - So far, a reliable workaround has been to merely deselect the element and reselect it.
- Keyman is the current default keyboard at the system level.
  - If another keyboard displays first, Keyman appears to have no issues when swapped in.

**To Reproduce**
Meet the conditions stated above in an attempt to use a keyboard search.

It doesn't seem to be 100% consistent - I _have_ seen a successful initial load before - but it seems to be at least 80% consistent, at least when I'm repeatedly attempting this via Simulator.

**Expected behavior**

Screen Shot 2020-11-27 at 1 41 34 PM

---

**iPhone/iPad:**
 - Keyman version 14.0.185

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #3985.  Likely by working around some sort of iOS bug... that I haven't found any online mention of.  Yay.

After identifying a 100% reproduction-rate sequence for #3985 (see [this comment](https://github.com/keymanapp/keyman/issues/3985#issuecomment-736274186) for details), I was finally able to start reliably debugging the issue.  After diving through the related call stacks, I noticed some _very_ peculiar behavior going on.

When the error occurs, it's because the system keyboard's constructors and `loadView` functions... were never called.  They're new and distinct instances, but our code never constructed them.  I thoroughly documented reference values and exact breakpoint locations to come to this conclusion, as it's a pretty serious accusation to make against iOS and/or the Swift compiler.

This matters because our keyboard makes the necessary calls to display the OSK... only after a successful load of the keyboard (see `keyboardLoaded`), which itself relies on `reloadKeyboard`, which is only called by `loadView`.  Since `loadView` is completely bypassed, the necessary event chain is bypassed, and the keyboard remains blank.

As a rough workaround, this solution implements a couple of simple checks in an attempt to detect and fix this scenario.  The checks aren't perfect, but I've written then in such a way that we'll only get "false negatives" that will result in an unnecessary reset.  You see, there's a race condition between the point where the keyboard's view appears and when the `resetKeyboard` callback (set within `keyboardLoaded`) is triggered to set the actual keyboard - this window of time will result in said 'false negative'.  While not optimal... since we don't know what mechanism is bypassing the constructors, and it "smells" like a deep-copy scenario... I'm not sure how to check for this window without introducing the possibility of false positives.

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html [hunk 1]
```diff
@@ -52,6 +52,15 @@
                     
                 //});
             }
+
+            function verifyLoaded() {
+              // During proper loads of KMW, a keyboard will be set very early on.
+              // There's a chance that the keyboard is still loading, so we double-check
+              // against the stub count.
+              if(!keyman.core.activeKeyboard && keyman.keyboardManager.keyboardStubs.length == 0) {
```
Comment (jahorton) [inline ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html:60]: The second part of this conditional aims to reduce "false negatives", preventing unnecessary resets.

**Comment by jahorton** [comment]
A few more notes about this weird scenario:

While the _constructors_ and _loadView_ are never called, the `viewWillAppear`, `viewDidAppear`, etc lifecycle methods _are_ still called and will respond to breakpoints in all cases - both when the original error occurs and during normal use.  Breakpoints in the constructors are completely bypassed _only_ when the error is reproduced using the documented repro sequence.

**Comment by jahorton** [comment]
Honestly, it's kind of shocking that the base load of the keyboard is actually available, despite the fact that our code never had the chance to pass the mysteriously-constructed new WebView the URL for the keyboard page.  Yet it's there.

**Comment by jahorton** [comment]
A little bit more ammo against Apple on this one - the break occurred after iOS 12.4, but before 13.3.  Did some testing across various Simulator runtimes; the issue simply doesn't reproduce in 12.4 or before, whereas it's 100% at and after 13.3.

**Comment by jahorton** [comment]
In case this may help others who run into the problem, in Swift terms that may turn up more easily in search engines:

Our `UIInputViewController`-derived class is somehow having its constructor bypassed, skipped, not called, etc in certain cases when used by our keyboard app-extension.  The _exact_ cause isn't known, but it's specific to recent versions of iOS and "smells" like a deep copy, despite the `coder` initializer being set with `fatalError` calls.

**Comment by mcdurdin** [comment]
> Honestly, it's kind of shocking that the base load of the keyboard page (`keyboard.html`) is actually there within the WebView, despite the fact that our code never had the chance to pass it the URL for the keyboard page. Yet it's there.

Weird. If the code is never called, how can the URL get there?

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift [hunk 2]
```diff
@@ -963,11 +966,24 @@ extension KeymanWebViewController {
...
 
     // Check for a change of "always show banner" state
     updateShowBannerSetting()
   }
 
+  /*
+   * Implemented as a workaround for a weird bug (likely within iOS) in which the
+   * InputViewController and KeymanWebViewController constructors (and thus, `loadView`)
+   * are sometimes completely bypassed during system-keyboard use.
+   */
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift:979]:
```suggestion
   * See https://github.com/keymanapp/keyman/issues/3985
   */
```

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html (hunk 1 above)
Comment (mcdurdin) [reply ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html:60]: Will this be an issue for 3rd party Keyman Engine consumers that may not have a default keyboard loaded?

**Review by mcdurdin — APPROVED** [review]

**Comment by jahorton** [comment]
> > Honestly, it's kind of shocking that the base load of the keyboard page (`keyboard.html`) is actually there within the WebView, despite the fact that our code never had the chance to pass it the URL for the keyboard page. Yet it's there.
> 
> Weird. If the code is never called, how can the URL get there?

Good question.  No, really, good question.  This is why I say that it "smells" like a deep copy scenario.

I could see a deep copy process noting "hey, what's the WKWebView's currently-loaded URL?", passing _that_ to the deep copy.  It would take extra effort to also deep-copy the results of all the possible effects that JS calls might have had, and I could see that being neglected.

**Comment by mcdurdin** [comment]
It's not a suspended process being restarted or anything like that? Can it happen on a clean install from scratch?

**Comment by jahorton** [comment]
> It's not a suspended process being restarted or anything like that? Can it happen on a clean install from scratch?

While I didn't fully erase all (Simulator) device settings, etc... every repro attempt I used to reach this conclusion was started by fully terminating the old processes and launching new ones via Xcode.  To my understanding, this means ripping out the old app installation - or at least the app-extension - and replacing it with the newly-compiled version.

I did reference tracking on each instance of `InputViewController` that was produced during a few repro tests.  The references were always distinct, and the old `WKWebView`s (from the older instances) were available for debugging in the Safari > Develop menu on the device, implying that they _shouldn't_ be a previously-frozen copy of anything.  I even traced a few `deinit` calls that resulted for the old `WKWebView` whenever a new `InputViewController` became active and built its own instance.

And, for good measure, I just now went ahead and did a full `Simulator` > `Device` > `Erase all Contents and Settings...` for an iPhone SE on 13.2.2.  I carried out the repro successfully as the first thing I did on the reset 'device'.

File: ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html (hunk 1 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html:60]: Hmm.  So, basically an iOS equivalent to https://github.com/keymanapp/keyman/issues/3922?

If no keyboard is installed, I _believe_ that `sil_eurolatin` will be automatically loaded in its place:

https://github.com/keymanapp/keyman/blob/0c015fe5c96000aeea2e9e995e7cb79e31e77134/ios/engine/KMEI/KeymanEngine/Classes/KeymanWebViewController.swift#L946-L961

(Note the final block.)

**Pull request merged**